--- arch/arm64/boot/dts/rockchip/rk356x.dtsi.org	2024-01-29 20:35:47.636423375 +0100
+++ arch/arm64/boot/dts/rockchip/rk356x.dtsi	2024-01-29 20:36:47.502566544 +0100
@@ -1524,17 +1524,17 @@
 
 			trips {
 				cpu_alert0: cpu_alert0 {
-					temperature = <85000>;
+					temperature = <95000>;
 					hysteresis = <2000>;
 					type = "passive";
 				};
 				cpu_alert1: cpu_alert1 {
-					temperature = <90000>;
+					temperature = <100000>;
 					hysteresis = <2000>;
 					type = "passive";
 				};
 				cpu_crit: cpu_crit {
-					temperature = <95000>;
+					temperature = <105000>;
 					hysteresis = <2000>;
 					type = "critical";
 				};
