--- arch/arm64/boot/dts/rockchip/rk3328-rock64.dts.org	2020-05-19 23:58:58.404096520 +0200
+++ arch/arm64/boot/dts/rockchip/rk3328-rock64.dts	2020-05-20 00:16:08.369331921 +0200
@@ -191,6 +191,26 @@
 	status = "okay";
 };
 
+/* add the clocks higher than 1.292ghz but disabled */
+&cpu0_opp_table {
+	opp-1392000000 {
+		opp-hz = /bits/ 64 <1392000000>;
+		opp-microvolt = <1350000>;
+		clock-latency-ns = <40000>;
+		status = "disabled";
+	};
+	opp-1512000000 {
+		opp-hz = /bits/ 64 <1512000000>;
+		opp-microvolt = <1450000>;
+		clock-latency-ns = <40000>;
+		status = "disabled";
+	};
+};
+
+&gpu {
+	status = "okay";
+};
+
 &hdmi {
 	status = "okay";
 
